examples/mcts.dx

```diff
@@ -0,0 +1,323 @@
+    maxResult = if leaf_min_player 
+                    then (1. - newResult)
+                    else newResult
+
+    -- collect itself and all ancestors
+    ancestors_list = yield_accum (ListMonoid node_ix) \ls.
+      with_state node \node_ref. iter \_.
+        append ls (get node_ref)
+        curr_node = from_just $ get (fst_ref tree_ref)!(get node_ref)
+        (Branch _ curr_parent _) = curr_node
+        if is_nothing curr_parent || (0@node_ix) == (get node_ref)
+          then Done ()
+          else
+            node_ref := from_just curr_parent
+            Continue
+
+    -- update win and loss counts
+    (AsList _ ancestors) = ancestors_list
+    for i.
+      -- unpack
+      ancestor = from_just $ get (fst_ref tree_ref)!(ancestors.i)
+      (Branch stats parent children) = ancestor
+      (wins, visits, level, state) = stats
+      -- compute
+      (new_wins, new_visits) = (wins + maxResult, visits + 1)
+      -- pack
+      new_stats = (new_wins, new_visits, level, state)
+      (fst_ref tree_ref)!(ancestors.i) := Just (Branch new_stats parent children)
+    ()
+
+' Wrap up the procedure for taking one sample and extend the tree.
+
+-- slow: every sample makes two copies of the tree array
+def mcts_sample {move state node_ix} [Eq node_ix]
+    (game: Game state move) (key: Key)
+    (tree: Tree node_ix (NodeStats state))
+    : Tree node_ix (NodeStats state) =
+    [key1, key2] = split_key key
+    leaf_ix = mcts_select game tree key1
+    leaf = from_just (fst tree).leaf_ix
+
+    tree' = yield_state tree \tree_ref.
+      if not $ (get_at #game_over game) $ node_state leaf then
+        mcts_expand game tree_ref leaf_ix
+
+    num_leaf_rollouts = 10
+    result = yield_accum (AddMonoid Float) \c.
+      for i:(Fin num_leaf_rollouts).
+        key' = ixkey key i
+        c += if mcts_rollout game key' $ node_state leaf 
+          then 1. / (n_to_f num_leaf_rollouts)
+          else 0.
+
+    yield_state tree' \tree_ref.
+      mcts_backpropagate game tree_ref leaf_ix result
+
+' ### Toy game: Count 21
+This is a simple two-player game. Two players in turn chooses
+between 1 and 2 to add to the current sum, starting at 0. Whoever reaches 21 wins.
+
+Count21State = Nat
+Count21Move = Nat
+Count21 : Game Count21State Count21Move = {
+  possible_moves = \sum:Count21State. to_list $ map unsafe_i_to_n [1,2],
+  win_or_lose = \sum. sum == unsafe_i_to_n 21,
+  update = \sum:Count21State move. sum + move,
+  game_over = \sum. sum >= unsafe_i_to_n 21,
+}
+
+' #### Approximate winning chances by sampling
+In theory, multiples of three are bad positions for the current player,
+since the opponent has a winning strategy by always choosing
+(3 - move by the current player). Otherwise, the current player
+has a winning strategy.
+
+key = new_key 0
+
+def test (key:Key) (current_state:Nat) : (Float & Nat) = 
+  init_node_stats : NodeStats Count21State = 
+    (0., unsafe_i_to_n 0, unsafe_i_to_n 0, current_state)
```

Comment:
The literal 0 should type-check as a `Nat` too?

Reply:
Done